File: electron/lib/kernel.js
/**
 * Hermes executable resolution + kernel (runtime) verification.
 * Extracted from main.js. Depends only on fs/os/path/crypto.
 */
const path = require('path')
const os = require('os')
const fs = require('fs')
const fsPromises = require('fs').promises
const crypto = require('crypto')

function resolveHermesCandidates() {
  const managedRoot = path.join(os.homedir(), 'AppData', 'Local', 'hermes', 'hermes-agent')
  // Prefer `venv` (the user's known-good, integration-patched runtime) first;
  // fall back to `.venv` (provisioned by newer `hermes update`). Both may exist.
  const cands = [
    path.join(managedRoot, 'venv', 'Scripts', 'hermes.exe'),
    path.join(managedRoot, '.venv', 'Scripts', 'hermes.exe'),
  ]
  try {
    const { execSync } = require('child_process')
    const out = execSync('where hermes 2>nul || which hermes 2>/dev/null').toString().trim()
    if (out) out.split(/\r?\n/).forEach(l => l.trim() && cands.push(l.trim()))
  } catch { /* ignore */ }
  const seen = new Set()
  const existing = []
  for (const c of cands) {
    if (seen.has(c)) continue
    seen.add(c)
    try { if (fs.existsSync(c)) existing.push(c) } catch { /* ignore */ }
  }
  return existing
}

// Resolve the preferred hermes executable (first existing candidate).
function resolveHermesCmd() {
  return resolveHermesCandidates()[0] || null
}

// ── Kernel verification (source path + Ed25519 signature) ────────────────────

function isTrustedPath(p) {
  // Kernel should live under known managed locations, not arbitrary paths.
  const localApp = process.env.LOCALAPPDATA || path.join(os.homedir(), 'AppData', 'Local')
  const trustedRoots = [
    path.join(localApp, 'hermes'),
    process.resourcesPath,
    path.dirname(process.execPath),
    process.cwd(),
  ]
  const rp = path.resolve(p)
  return trustedRoots.some(root => rp.startsWith(path.resolve(root)))
}

async function sha256File(filePath) {
  const data = await fsPromises.readFile(filePath)
  return crypto.createHash('sha256').update(data).digest('hex')
}

async function listKernelArtifacts(hermesCmdPath) {
  const artifacts = []
  if (!hermesCmdPath || !fs.existsSync(hermesCmdPath)) return artifacts
  artifacts.push({ id: 'entry', path: hermesCmdPath, hash: await sha256File(hermesCmdPath) })
  const baseDir = path.dirname(hermesCmdPath)
  const candidates = [
    path.join(baseDir, 'hermes'),
    path.join(baseDir, 'hermes-cli'),
    path.join(baseDir, 'hermes_cli'),
    path.join(baseDir, 'python.exe'),
    path.join(baseDir, '..', 'Lib', 'site-packages', 'hermes', '__init__.py'),
  ]
  for (const c of candidates) {
    try {
      if (fs.existsSync(c) && fs.statSync(c).isFile()) {
        artifacts.push({ id: path.relative(baseDir, c), path: c, hash: await sha256File(c) })
      }
    } catch {}
  }
  return artifacts
}

async function loadKernelPublicKey() {
  const candidates = [
    path.join(process.resourcesPath, 'kernel.pub'),
    path.join(process.resourcesPath, 'assets', 'kernel.pub'),
    path.join(os.homedir(), 'AppData', 'Local', 'hermes', 'kernel.pub'),
    path.join(__dirname, '..', 'kernel.pub'),
  ]
  for (const c of candidates) {
    try {
      if (fs.existsSync(c)) return await fsPromises.readFile(c)
    } catch {}
  }
  return null
}

async function verifyKernelSignature(hermesCmdPath) {
  const pubKey = await loadKernelPublicKey()
  if (!pubKey) {
    return { ok: false, hasKey: false, message: '未包含官方公钥（开发构建），已跳过 Ed25519 校验' }
  }
  const sigPath = hermesCmdPath + '.sig'
  if (!fs.existsSync(sigPath)) {
    return { ok: false, hasKey: true, hasSig: false, message: '未找到运行时签名文件 ' + sigPath }
  }
  try {
    const data = await fsPromises.readFile(hermesCmdPath)
    const sig = await fsPromises.readFile(sigPath)
    const ok = crypto.verify(null, data, pubKey, sig)
    return { ok, hasKey: true, hasSig: true, message: ok ? 'Ed25519 签名校验通过' : 'Ed25519 签名校验失败' }
  } catch (e) {
    return { ok: false, hasKey: true, hasSig: true, message: '签名校验出错：' + (e && e.message) }
  }
}

async function verifyKernel() {
  const hermesCmdPath = resolveHermesCmd()
  if (!hermesCmdPath) {
    return { ok: false, status: 'unknown', message: '未找到 Hermes 运行时可执行文件', artifacts: [], combinedHash: '' }
  }
  if (!isTrustedPath(hermesCmdPath)) {
    return { ok: false, status: 'untrusted', message: '运行时路径不在受信任安装目录中：' + hermesCmdPath, artifacts: [], combinedHash: '' }
  }
  const artifacts = await listKernelArtifacts(hermesCmdPath)
  const sigResult = await verifyKernelSignature(hermesCmdPath)
  const integrityInput = artifacts.map(a => a.hash).join('')
  const combinedHash = crypto.createHash('sha256').update(integrityInput).digest('hex').slice(0, 32)
  const status = sigResult.ok ? 'verified' : 'unverified'
  const message = sigResult.ok
    ? `内核来源已校验，完整性哈希 ${combinedHash}`
    : `${sigResult.message}；完整性哈希 ${combinedHash}`
  return { ok: sigResult.ok, status, message, artifacts, combinedHash, sig: sigResult }
}

module.exports = {
  resolveHermesCandidates,
  resolveHermesCmd,
  isTrustedPath,
  sha256File,
  listKernelArtifacts,
  loadKernelPublicKey,
  verifyKernelSignature,
  verifyKernel,
}

File: src-tauri/src/kernel.rs
//! Hermes executable resolution + kernel (runtime) verification.
//! Port of `electron/lib/kernel.js`.

use crate::paths::{hermes_agent_dir, venv_hermes_bin};
use ed25519_dalek::{Signature, SignatureError, Verifier, VerifyingKey};
use serde::Serialize;
use sha2::{Digest, Sha256};
use std::path::{Path, PathBuf};
use std::process::Command;

/// Ordered list of candidate hermes executables, deduplicated, existing only.
pub fn resolve_hermes_candidates() -> Vec<PathBuf> {
    let mut cands: Vec<PathBuf> = vec![
        venv_hermes_bin(Some(&hermes_agent_dir()), "venv"),
        venv_hermes_bin(Some(&hermes_agent_dir()), ".venv"),
    ];
    // A `hermes` / `hermes-agent` on PATH works as a last resort.
    if let Ok(out) = Command::new("sh").arg("-lc").arg("command -v hermes hermes-agent 2>/dev/null").output() {
        if out.status.success() {
            let s = String::from_utf8_lossy(&out.stdout);
            for l in s.lines() {
                let l = l.trim();
                if !l.is_empty() {
                    cands.push(PathBuf::from(l));
                }
            }
        }
    }
    let mut seen = std::collections::HashSet::new();
    let mut existing = Vec::new();
    for c in cands {
        if seen.insert(c.clone()) && c.exists() {
            existing.push(c);
        }
    }
    existing
}

/// Preferred hermes executable (first existing candidate), or None.
pub fn resolve_hermes_cmd() -> Option<PathBuf> {
    resolve_hermes_candidates().into_iter().next()
}

fn is_trusted_path(p: &Path) -> bool {
    let hermes_dir = crate::paths::hermes_data_dir();
    let trusted_roots = vec![
        hermes_dir,
        dirs::home_dir().unwrap_or_else(|| PathBuf::from("/")),
    ];
    if let Ok(rp) = std::fs::canonicalize(p) {
        for root in trusted_roots {
            if let Ok(rr) = std::fs::canonicalize(&root) {
                if rp.starts_with(&rr) {
                    return true;
                }
            }
        }
    }
    false
}

fn sha256_file(path: &Path) -> Option<String> {
    std::fs::read(path).ok().map(|d| hex::encode(Sha256::digest(&d)))
}

#[derive(Serialize, Default)]
pub struct KernelSig {
    pub ok: bool,
    pub has_key: bool,
    pub has_sig: bool,
    pub message: String,
}

#[derive(Serialize, Default)]
pub struct KernelVerify {
    pub ok: bool,
    pub status: String,
    pub message: String,
    pub combined_hash: String,
    pub sig: KernelSig,
}

fn load_kernel_public_key() -> Option<Vec<u8>> {
    for c in [
        crate::paths::hermes_data_dir().join("kernel.pub"),
        dirs::home_dir()
            .unwrap_or_default()
            .join(".local")
            .join("share")
            .join("hermes")
            .join("kernel.pub"),
    ] {
        if c.exists() {
            if let Ok(b) = std::fs::read(&c) {
                return Some(b);
            }
        }
    }
    None
}

fn verify_ed25519(data: &[u8], sig: &[u8], pubkey: &[u8]) -> Result<bool, SignatureError> {
    let key = VerifyingKey::from_bytes(pubkey.try_into().map_err(|_| SignatureError::from_source("bad key len"))?)?;
    let sig_obj = Signature::from_bytes(sig.try_into().map_err(|_| SignatureError::from_source("bad sig len"))?);
    Ok(key.verify(data, &sig_obj).is_ok())
}

pub fn verify_kernel() -> KernelVerify {
    let mut out = KernelVerify {
        ok: false,
        status: "unknown".into(),
        message: String::new(),
        sig: KernelSig::default(),
        combined_hash: String::new(),
    };
    let Some(cmd) = resolve_hermes_cmd() else {
        out.status = "unknown".into();
        out.message = "未找到 Hermes 运行时可执行文件".into();
        return out;
    };
    if !is_trusted_path(&cmd) {
        out.status = "untrusted".into();
        out.message = format!("运行时路径不在受信任安装目录中: {}", cmd.display());
        return out;
    }
    let artifacts: Vec<(&str, String)> = {
        let mut v = Vec::new();
        if let Some(h) = sha256_file(&cmd) {
            v.push(("entry", h));
        }
        let base = cmd.parent().unwrap_or(Path::new("/"));
        for (id, cand) in [
            ("hermes", base.join("hermes")),
            ("python", if cfg!(windows) { base.join("python.exe") } else { base.join("python") }),
        ] {
            if cand.is_file() {
                if let Some(h) = sha256_file(&cand) {
                    v.push((id, h));
                }
            }
        }
        v
    };
    let combined = {
        let mut h = Sha256::new();
        for (_, hash) in &artifacts {
            h.update(hash.as_bytes());
        }
        hex::encode(h.finalize())[..32].to_string()
    };
    out.combined_hash = combined.clone();

    let mut sig = KernelSig::default();
    let pubkey = load_kernel_public_key();
    match pubkey {
        None => {
            sig.ok = false;
            sig.has_key = false;
            sig.message = "未包含官方公钥（开发构建），已跳过 Ed25519 校验".into();
        }
        Some(pubkey) => {
            sig.has_key = true;
            let sig_path = PathBuf::from(format!("{}.sig", cmd.display()));
            if !sig_path.exists() {
                sig.has_sig = false;
                sig.message = format!("未找到运行时签名文件 {}", sig_path.display());
            } else {
                sig.has_sig = true;
                let data = std::fs::read(&cmd).unwrap_or_default();
                let sig_bytes = std::fs::read(&sig_path).unwrap_or_default();
                match verify_ed25519(&data, &sig_bytes, &pubkey) {
                    Ok(true) => {
                        sig.ok = true;
                        sig.message = "Ed25519 签名校验通过".into();
                    }
                    Ok(false) => {
                        sig.message = "Ed25519 签名校验失败".into();
                    }
                    Err(e) => {
                        sig.message = format!("签名校验出错: {e}");
                    }
                }
            }
        }
    }
    out.sig = sig;
    out.status = if out.sig.ok { "verified".into() } else { "unverified".into() };
    out.ok = out.sig.ok;
    out.message = if out.sig.ok {
        format!("内核来源已校验，完整性哈希 {combined}")
    } else {
        format!("{}；完整性哈希 {combined}", out.sig.message)
    };
    out
}
